parsers/GT.py

```diff
@@ -1,163 +1,206 @@
-                "solar": row["IRRADIACIÓN"],
-                "unknown": row["BIOMASA/CARBÓN"]
-                + row["CARBÓN/PETCOKE"]
-                + row["SYNGAN"],
-                "wind": row["VIENTO"],
-            },
-            "source": DOMAIN,
-            "zoneKey": zone_key,
-        }
-        for hour, row in enumerate(api_data)
-    ]
-    return [
-        validation.validate(
-            result, logger, floor=PRODUCTION_THRESHOLD, remove_negative=True
-        )
-        for result in results
-    ]
-
 
-def _get_api_data(
-    session: Session | None,
-    url: str,
-    date_time: datetime,
-    target: ApiKind,
-) -> list[dict]:
-    """Get the JSON-formatted response from the AMM API for the desired date-time."""
-    session = session or Session()
-    response_payload = session.get(
-        url, params={"dt": date_time.strftime("%d/%m/%Y")}
-    ).json()
-
-    # The JSON data returned by the API is a list of objects, each
-    # representing one technology type. Collect this information into a list,
-    # with the list index representing the hour of day.
-    results = [collections.defaultdict(float) for _ in range(24)]
-    for row in response_payload:
-        # The API returns hours in the range [1, 24], so each one refers to the
-        # past hour (e.g., 1 is the time period [00:00, 01:00)). Shift the hour
-        # so each index represents the hour ahead and is in the range [0, 24),
-        # e.g., hour 0 represents the period [00:00, 01:00).
-        results[int(row["hora"]) - 1][row["tipo"]] = row["potencia"]
-
-    is_historical_data = date_time.date() < datetime.now(TIMEZONE).date()
-
-    # For live consumption data, an hour's data isn't updated until the hour has passed,
-    # so the current (and future) hour(s) should not be included in the results.
-    # For live production data, the API will return zero-filled future data until the end of the day,
-    # so future hours should not be included in the results.
-    cutoff_index = (
-        date_time.hour if target == ApiKind.CONSUMPTION else date_time.hour + 1
-    )
+    production_breakdown_list = ProductionBreakdownList(logger)
+    for dt, row in api_data.items():
+        production_mix = ProductionMix()
+        production_mix.add_value("biomass", row["BIOGAS"] + row["BIOMASA"])
+        production_mix.add_value("coal", row["CARBÓN"])
+        production_mix.add_value("gas", row["GAS NATURAL"])
+        production_mix.add_value("geothermal", row["VAPOR"])
+        production_mix.add_value("hydro", row["AGUA"])
+        production_mix.add_value("oil", row["BUNKER"] + row["DIESEL"])
+        production_mix.add_value("solar", row["IRRADIACIÓN"])
+        production_mix.add_value(
+            "unknown",
+            # 'BIOMASA/CARBÓN not always present, e.g. 15/07/2023 TZ
+            row.get("BIOMASA/CARBÓN", 0.0) + row["CARBÓN/PETCOKE"] + row["SYNGAN"],
+        )
+        production_mix.add_value("wind", row["VIENTO"])
```

Comment:
I would prefer if this was done as a loop instead with a mapping to the maps modes. That way we can avoid the get.("...", 0.0) as we really don't want default values to be set in the parsers.

By doing it in a loop we can also get more accurate floating point correction as it's done after each addition instead of on the total. This should probably not be a major issue but still.

(it's still there in some legacy parsers but should be removed whenever possible).

There looks to be some remapping that should be done here as well from the unknown to known modes but I'll address that later.